[FX] Pickle serialization of GraphModule via forward source

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43741 [FX][RFC] Input prototype for collection types in sym trace
* #43740 [FX] Better error when unpacking Proxy
* **#43674 [FX] Pickle serialization of GraphModule via forward source**
* #43666 [FX] Support tensor-valued constants

More of an RFC at this point, but here I define `__reduce__` on `GraphModule` to use the generated source as a way to transmit the code of the `Module`

Closes https://github.com/pytorch/pytorch/issues/43513

Differential Revision: [D23362396](https://our.internmc.facebook.com/intern/diff/D23362396)

Commit:
[FX] Pickle serialization of GraphModule via forward source

[ghstack-poisoned]

Commit:
Update on "[FX] Pickle serialization of GraphModule via forward source"


More of an RFC at this point, but here I define `__reduce__` on `GraphModule` to use the generated source as a way to transmit the code of the `Module`

Closes https://github.com/pytorch/pytorch/issues/43513

Differential Revision: [D23362396](https://our.internmc.facebook.com/intern/diff/D23362396)

[ghstack-poisoned]

dzhulgakov commented on torch/fx/graph_module.py:
```diff
@@ -26,6 +27,24 @@ def patched_getline(*args, **kwargs):
+    }
+    exec_with_source(src, gbls)
+    cls = type(root)
+    for k, v in gbls.items():
```
I think it sets an attribute called `torch` on every module :) I think the same is true below too

jamesr66a replied:
It also dumps `__builtins__` onto the Module for some reason

dzhulgakov commented on torch/fx/graph_module.py:
```diff
@@ -26,6 +27,24 @@ def patched_getline(*args, **kwargs):
+    to the original code. Then it symbolically traces through it to get the
+    GraphModule
+    """
+    root = copy.deepcopy(root)
```
why do you need deepcopy here? Isn't the original module just created by pickle?

jamesr66a replied:
It is now, but inevitably somebody is going to find this function and use it for nefarious purposes, so this is trying to be defensive. It would be surprising for this function to both mutate its argument and return an entirely new Module instance

jamesr66a replied:
I can just change it to a shallow copy

dzhulgakov replied:
if you get rid of `root` then it won't mutate the argument, right? I.e. `symbolic_trace` will just create new subhierarchy every time only from the submodules that are used. And what mutation are we talking about? Is it just the "tensor constants"? that should be address with `root` elimination and I don't think there are any other mutations.

also if you don't want it to be discoverable - put `_` in front. I find it wasteful to put an extra deepcopy in unpickling path as it's absolutely not needed

jamesr66a replied:
The mutation I'm referring to is line 46, where we're dumping all of the globals onto the passed-in `root` module. For any arbitrary module that's passed in, it's going to (at the very least) wipe out any existing forward method that's on it

Re: adding a leading underscore, I see that as more of a Python cargo cult. It's not enforced and people use our internal APIs all the time.

jamesr66a replied:
> if you get rid of root

We haven't done that :p

dzhulgakov commented on torch/fx/graph_module.py:
```diff
@@ -66,6 +85,9 @@ def forward(self, {', '.join(free_variables)}):
         for k, v in gbls.items():
```
I think the same bug is here too

jamesr66a replied:
I guess the question is if there's any name other than `forward` we want to extract from the globals after `exec`. @zdevito  do you know?

dzhulgakov replied:
This can be done in a separate PR, so non-blocking

dzhulgakov reviewed (commented):
Before we do this - can we sort out whether we want to eliminate `.root` reference and assign submodules directly on the parent one? because that will change what you need to do here a bit

Commit:
Update on "[FX] Pickle serialization of GraphModule via forward source"


More of an RFC at this point, but here I define `__reduce__` on `GraphModule` to use the generated source as a way to transmit the code of the `Module`

Closes https://github.com/pytorch/pytorch/issues/43513

Differential Revision: [D23362396](https://our.internmc.facebook.com/intern/diff/D23362396)

[ghstack-poisoned]

Commit:
Update on "[FX] Pickle serialization of GraphModule via forward source"


More of an RFC at this point, but here I define `__reduce__` on `GraphModule` to use the generated source as a way to transmit the code of the `Module`

Closes https://github.com/pytorch/pytorch/issues/43513

Differential Revision: [D23362396](https://our.internmc.facebook.com/intern/diff/D23362396)

[ghstack-poisoned]

dzhulgakov reviewed (approved):
Please remove deepcopy, otherwise it's good to go

Commit:
Update on "[FX] Pickle serialization of GraphModule via forward source"


More of an RFC at this point, but here I define `__reduce__` on `GraphModule` to use the generated source as a way to transmit the code of the `Module`

Closes https://github.com/pytorch/pytorch/issues/43513

Differential Revision: [D23362396](https://our.internmc.facebook.com/intern/diff/D23362396)

[ghstack-poisoned]

bradleyhd commented:
@jamesr66a will this enable pickling of custom ops in the graph? currently having issues pickling graphs when the target of a node is a method of a `PyCapsule` object and not sure what approaches to take to unblock

jamesr66a commented:
@bradleyhd if you're encountering issues with pickling PyCapsule objects, I believe that means you have a pybind11 object in there somewhere, which to be clear is not a custom op and is not supported. It would be easier if you could provide a repro

Commit:
Update on "[FX] Pickle serialization of GraphModule via forward source"


More of an RFC at this point, but here I define `__reduce__` on `GraphModule` to use the generated source as a way to transmit the code of the `Module`

Closes https://github.com/pytorch/pytorch/issues/43513

Differential Revision: [D23362396](https://our.internmc.facebook.com/intern/diff/D23362396)

[ghstack-poisoned]

Commit:
Update on "[FX] Pickle serialization of GraphModule via forward source"


More of an RFC at this point, but here I define `__reduce__` on `GraphModule` to use the generated source as a way to transmit the code of the `Module`

Closes https://github.com/pytorch/pytorch/issues/43513

Differential Revision: [D23362396](https://our.internmc.facebook.com/intern/diff/D23362396)

[ghstack-poisoned]
